File: 1335-minimum-difficulty-of-a-job-schedule/1335-minimum-difficulty-of-a-job-schedule.cpp
class Solution {
public:
    
    int minDifficulty(vector<int>& jobDifficulty, int days) {
        int n = jobDifficulty.size();
        if (n < days) {
            return -1;
        }
    
        vector<long long> dp(n + 1, INT_MAX);
        dp[n] = 0;
        for (int d = 1; d <= days; d++) {
            for (int i = 0; i <= n - d; i++) {
                int maxd = 0;
                dp[i] = INT_MAX;
                for (int j = i; j <= n - d; j++) {
                    maxd =  max(maxd, jobDifficulty[j]);
                    dp[i] = min(dp[i], maxd + dp[j + 1]);
                }
            }
        }
        return (int) dp[0];
    }
};

File: 1335-minimum-difficulty-of-a-job-schedule/1335-minimum-difficulty-of-a-job-schedule.java
class Solution {
    public int minDifficulty(int[] jobDifficulty, int days) {
        int len = jobDifficulty.length;
        if (len < days) {
            return -1;
        }
        int[] dp = new int[len + 1];
        Arrays.fill(dp, 1000000000);
        dp[len] = 0;
        for (int d = 1; d <= days; d++) {
            for (int i = 0; i <= len - d; i++) {
                int maxd = 0;
                dp[i] = 1000000000;
                for (int j = i; j <= len - d; j++) {
                    maxd =  Math.max(maxd, jobDifficulty[j]);
                    dp[i] = Math.min(dp[i], maxd + dp[j + 1]);
                }
            }
        }
        return dp[0];
    }
}